Generalize constitutive law with optional tensor

jperryhouts opened the pull request:
Still a work in progress.

Is backwards compatible with isotropic material models: per Timo's suggestion the original viscosity paramter is still used and the optional viscosity tensor is set to the identity by default.

Also, could someone clarify to me how the divergence of u term for the compressible case works in the bilinear form? Does it need to get multiplied by the trace of the viscosity tensor? I think this happens implicitly when viscosity is isotropic, but I've managed to confuse myself about it.

tjhei commented on include/aspect/material_model/interface.h:
```diff
@@ -824,6 +835,20 @@ namespace aspect
+         * computed without the $\frac 13 \nabla \cdot \mathbf u \mathbf 1$
+         * term as is required for consistency in the compressible case.
+         */
+        virtual SymmetricTensor<4,dim> viscosity_tensor (const double                  temperature,
```
I don't think we should add this function to ```InterfaceCompatibility```

jperryhouts replied:
Ok. Is that because using the InterfaceCompatibility class is depreciated, or for another reason?

tjhei replied:
It was introduced for backwards compatibility, yes. No reason to implement new features there. Instead, we should move to the newer interface.

tjhei commented:
One simplification would be to fill the outputs with the identity tensor by default. That way, you don't need ```is_anisotropic```.

Not sure about the div (rho u) term. I guess you need to go through the derivation again.

bangerth commented:
I think the patch needs to document somewhere that the viscosity that will be used is `eta` *times* the `viscosity_tensor`, not one or the other. In other words, one could consider the `viscosity_tensor` as a *direction* with unit length.

In your question about the div term, I suppose you mean the `pressure_scaling` thing? If so, yes, the pressure scaling should then include something about the viscosity tensor. The scaling only makes sure that the two equations are balanced, so the exact scaling constant is not important. This is currently implemented in line 387 in core.cc:
```
    pressure_scaling = material_model->reference_viscosity() / geometry_model->length_scale();
```
My suggestion would simply be to document that the `reference_viscosity()` function should take into account the (scalar) size of `eta` *and* the viscosity tensor. It could take the norm of the latter, or the Frobenius norm, or the largest entry, etc. What you choose does not matter very much.

jperryhouts commented:
@tjhei Would filling the vector by default be slow? I don't actually know if the dealii::identity_tensor() function needs to calculate anything, or if it just returns the hard coded value.

@bangerth Actually I think you're a step ahead of me, I was just hung up on how a viscosity tensor should multiply the the -1/3*eta*div(u) term in the bilinear form (and frankly, where that comes from at all). But as Timo mentioned I should probably just go back to the derivation, at least as a refresher for my own sake.

bangerth commented:
On 05/11/2015 05:37 PM, Jonathan Perry-Houts wrote:
> @bangerth <https://github.com/bangerth> Actually I think you're a step
> ahead of me, I was just hung up on how a viscosity tensor should
> multiply the the -1/3/eta/div(u) term in the bilinear form (and frankly,
> where that comes from at all). But as Timo mentioned I should probably
> just go back to the derivation, at least as a refresher for my own sake.

Yes. I think it comes from the definition of the compressible strain,
   1/2(grad u + grad u^T) - 1/3 (div u) I
which you then proceed to multiply with the viscosity to get the (shear) 
stress. The div u term has the identity matrix, so that's what you 
probably need to multiply by the fourth-order viscosity tensor then.

But yes, better to check :-)

jperryhouts commented:
@bangerth What do you think about this? I tried to make minimal changes, and keep the changes very general. The constitutive tensor variable could be used for other applications besides anisotropic viscosity, and doesn't change any functionality if it isn't assigned a value by a material model.

I split the material models I'm working on (for anisotropic viscosity, specifically) into another branch, to be considered in a future PR.

bangerth commented:
/run-tests

bangerth commented on include/aspect/material_model/interface.h:
```diff
@@ -222,6 +222,25 @@ namespace aspect
+       * caution when implementing compressible material models which use the
+       * constitutive tensor.
+       */
+      std::vector<SymmetricTensor<4,dim> > consitutive_tensors;
```
there's a misspelling in the name of the variable that you probably want to fix throughout.

bangerth commented on include/aspect/material_model/interface.h:
```diff
@@ -222,6 +222,25 @@ namespace aspect
+       *
+       * @note The strain rate term in equation (1) of the manual will be
+       * multiplied by this tensor *and* the viscosity scalar (\\eta). If no
+       * value is assigned here the identity tensor will be used by default.
```
Can you explain what "identity tensor" here actually means? I suppose you mean the rank-4 tensor `I` that when multiplied by a symmetric rank-2 tensor `epsilon` has the property that `I epsilon = epsilon`. That is the case for `I_{ijkl} = \delta_{ik}\delta_{jl}`.

bangerth commented on include/aspect/material_model/interface.h:
```diff
@@ -222,6 +222,25 @@ namespace aspect
+       * multiplied by this tensor *and* the viscosity scalar (\\eta). If no
+       * value is assigned here the identity tensor will be used by default.
+       *
+       * @note because the use of this variable generalizes the constitutive
```
capitalize `because`

bangerth commented on include/aspect/material_model/interface.h:
```diff
@@ -222,6 +222,25 @@ namespace aspect
+       * manual. It is therefore only multiplied by the strain rate term in
+       * the governing equations and left up to the user to proceed with
+       * caution when implementing compressible material models which use the
+       * constitutive tensor.
```
I think this calls for a section in the manual that discusses in more detail how exactly this tensor is used, including the formulas where it is explicitly shown where the tensor appears and where not.

bangerth commented on include/aspect/material_model/interface.h:
```diff
@@ -585,6 +604,13 @@ namespace aspect
          * the pressure against the velocity. It is also used in computing
          * dimension-less quantities.
+         *
+         * @note the reference viscosity should take in to account the complete
```
capitalize `the`

bangerth replied:
in to -> into

bangerth commented on source/simulator/assembly.cc:
```diff
@@ -1014,6 +1014,9 @@ namespace aspect
         const double eta = scratch.material_model_outputs.viscosities[q];
 
+        const SymmetricTensor<4,dim> constitutive_tensor =
+          scratch.material_model_outputs.consitutive_tensors[q];
```
I'd suggest making this a reference, rather than just copy the tensor. A rank-4 tensor in dim=3 has 6^2=36 elements, which requires 288 bytes to be copied. I think a reference may be the better way to do this.

bangerth commented on source/simulator/assembly.cc:
```diff
@@ -1214,12 +1218,15 @@ namespace aspect
                             :
                             std::numeric_limits<double>::quiet_NaN());
 
+        const SymmetricTensor<4,dim> constitutive_tensor = scratch.material_model_outputs.consitutive_tensors[q];
```
same here

bangerth commented:
I thin what this patch requires is (i) a short section in the manual, say after 2.7, that explains in detail where this tensor appears in the equations (and, if necessary, where it doesn't), and (ii) I'd be more comfortable if you could run a couple of timesteps of a 3d testcase (both in debug and in release mode) where you compare the assembly time with and without your patch; I suspect they will not differ by very much, but I'd like to be sure.

jperryhouts commented:
Sounds good. I'll get to this as soon as I have a chance. I'm teaching field camp out in Montana this month so my computer time will be limited. The thing about speed worries me a bit, too. I am interested to see what the actual effect is though.

jperryhouts commented:
@bangerth You were right, all the tensor multiplication slowed the code down by a factor of ~2x in release mode and ~13x in debug mode (I guess dealii has lots of asserts in the tensor multiplication function?).

I fixed it, and the default behavior is back to un-resolvably different runtimes from the pre-patched code, but the solution looks a bit ugly. Let me know what you think, and specifically if there is any way to do this more elegantly.

jperryhouts commented:
Actually, debug mode still runs ~2x slower with the patch than without. Not sure why yet.

jperryhouts commented:
Nope, scratch that. Runtimes are the same. Sorry for the spam!!

jperryhouts commented:
Ok. I, somewhat cautiously, think that this patch is finally ready to go. I don't think I can make it any less obtrusive (although I'm not totally happy with the way it looks) but it works and doesn't sacrifice efficiency in the default configuration. Comments, @bangerth and @tjhei ?

jperryhouts commented:
Actually, I still need to tie up loose ends with compressibility and this constitutive law implementation. Comments are still welcome on everything else, though.

tjhei commented:
Looks very good. Can you please add a test case that uses this new tensor?

jperryhouts commented:
OK, I fixed the compressible case, made the shear heating model consistent with my changes, described the whole thing in the manual, and added a test which uses a constitutive tensor, in a compressible model, with shear heating.

So long as the tester is okay with all that, i think this is actually *finally* in a state where I wouldn't be embarrassed to merge it!!

gassmoeller commented:
Hi Jonathan, thanks a lot, this looks nearly good to go. Last question: Is it possible to make the testcase somewhat smaller (reduce refinements by 2) or omit the visualization output and output velocity statistics instead? This would save us 7000 lines of test output that might be updated now and then ;-)

jperryhouts commented:
Yeah, if that's something we're worried about, I think the statistics file is *probably* enough to be diagnostic. I updated the PR with that change.

tjhei commented:
Looks good, @bangerth are you also happy with this interface change?

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1138,6 +1138,46 @@ \subsection{Compositional fields}
+\subsection{Constitutive laws}
+
+Equation (\ref{eq:stokes-1}) describes buoyancy-driven flow in an isotropic
+fluid where strain rate is related to stress by a scalar field, $\eta$. For
```
I think it would be clearer to say "by a scalar (possibly spatially variable) multiplier, $\eta$".

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1138,6 +1138,46 @@ \subsection{Compositional fields}
+Equation (\ref{eq:stokes-1}) describes buoyancy-driven flow in an isotropic
+fluid where strain rate is related to stress by a scalar field, $\eta$. For
+some material models it is useful to generalize this relationship for anisotropic
+materials, or other exotic constitutive laws. For these cases \aspect{} can
```
for -> to anisotropic materials

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1138,6 +1138,46 @@ \subsection{Compositional fields}
+\end{align}
+where $C = C_{ijkl}$ is defined by the material model. This functionality can be
+optionally invoked by any material model that chooses to define a $C$ field, and
+falls back to the default case ($C=\mathbf 1$) if no such field is defined. It
```
I think for fourth order tensors, one typically denotes the identity tensor by `\mathbb I`.

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1138,6 +1138,46 @@ \subsection{Compositional fields}
+optionally invoked by any material model that chooses to define a $C$ field, and
+falls back to the default case ($C=\mathbf 1$) if no such field is defined. It
+should be noted that $\eta$ still appears in equations (\ref{eq:stokes-1-anisotropic})
+and (\ref{eq:temperature-anisotropic}). $C$ is therefore intended to be thought of
```
make your life simpler and use `\eqref` instead of `(\ref...)` :-)

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1138,6 +1138,46 @@ \subsection{Compositional fields}
+  \dots
+  \notag
+\end{align}
+where $C = C_{ijkl}$ is defined by the material model. This functionality can be
```
Add a sentence here such as this:

`For physical reasons, $C$ needs to be a symmetric rank-4 tensor, i.e., when multiplied by a symmetric (strain rate) tensor of rank 2, it needs to return another symmetric tensor of rank 2. In mathematical terms, this means that $C_{ijkl}=C_{jikl}=C_{ijlk}=C_{jilk}$. Energy considerations also require that $C$ is positive definite, i.e., for any $\varepsilon\neq 0$, the scalar $\varepsilon : (C \varepsilon)$ must be positive.`

bangerth commented on include/aspect/material_model/interface.h:
```diff
@@ -274,6 +274,18 @@ namespace aspect
+       * viscosity.
+       *
+       * @note The strain rate term in equation (1) of the manual will be
+       * multiplied by this tensor *and* the viscosity scalar (\\eta). If no
```
\\eta -> `$\\eta`

bangerth commented on include/aspect/material_model/interface.h:
```diff
@@ -274,6 +274,18 @@ namespace aspect
+       * value is assigned here the identity tensor will be used in its place.
+       * This leaves the isotropic constitutive law unchanged by default.
+       */
+      std::vector<SymmetricTensor<4,dim> > constitutive_tensors;
```
I find the term `constitutive_tensors` too vague. How about `stress_strain_directors` to match the terminology you use in the manual? You could also reference the manual here.

bangerth replied:
By the way, what will this variable be set to "if no value is assigned here"? Does that mean the vector is simply empty, or that it has the right size and simply has special values?

bangerth commented on source/heating_model/shear_heating.cc:
```diff
@@ -41,15 +41,30 @@ namespace aspect
-          heating_model_outputs.heating_source_terms[q] = 2.0 * material_model_outputs.viscosities[q] *
-                                                          compressible_strain_rate * compressible_strain_rate;
+          const SymmetricTensor<4,dim> C = material_model_outputs.constitutive_tensors[q];
+          const SymmetricTensor<2,dim> directed_strain_rate = ((C == dealii::identity_tensor<dim> ())
```
This is a potentially expensive comparison: it requires comparing all 36 elements of the two tensors.

If you assume that the elements of the constitutive_tensors array are set to anything at all (which you do here), then I'd simply always do the multiplication. Loading the data from memory so you can do the comparison will in practice not take more time than actually doing the multiplication. But it will keep the code easier to read.

jperryhouts replied:
I agree, this is awkward to read. And maybe it's some compiler magic, but doing the comparison seems to make no difference in assembly time, and doing the multiplication slows down the entire model by an order of magnitude. Since I'm just using the identity tensor as a sort of "signaling" null value, is there a more obvious object to use in its place? If it doesn't make a difference, I kinda like using the identity tensor because it's sort of self-documenting in a way.

bangerth commented on source/heating_model/shear_heating.cc:
```diff
@@ -41,15 +41,30 @@ namespace aspect
-
-          heating_model_outputs.heating_source_terms[q] = 2.0 * material_model_outputs.viscosities[q] *
-                                                          compressible_strain_rate * compressible_strain_rate;
+          const SymmetricTensor<4,dim> C = material_model_outputs.constitutive_tensors[q];
```
Since this object is large, make `C` a reference.

bangerth commented on source/simulator/assembly.cc:
```diff
@@ -997,14 +997,19 @@ namespace aspect
 
+        const SymmetricTensor<4,dim> &constitutive_tensor =
+          scratch.material_model_outputs.constitutive_tensors[q];
+        const bool use_tensor = (constitutive_tensor != dealii::identity_tensor<dim> ());
```
same here -- I'd simply *always* do the multiplication with the tensor.

bangerth commented on source/simulator/assembly.cc:
```diff
@@ -1197,12 +1202,17 @@ namespace aspect
 
+        const SymmetricTensor<4,dim> &constitutive_tensor =
+          scratch.material_model_outputs.constitutive_tensors[q];
+        const bool use_tensor = (constitutive_tensor !=  dealii::identity_tensor<dim> ());
```
and here

bangerth commented on tests/anisotropic_viscosity.cc:
```diff
@@ -0,0 +1,161 @@
+          const unsigned int vi3d0[] = {0, 1, 2, 1, 0, 0};
+          const unsigned int vi3d1[] = {0, 1, 2, 2, 2, 1};
+          const unsigned int vi2d0[] = {0, 1, 0};
+          const unsigned int vi2d1[] = {0, 1, 1};
```
Ah, cool -- I think I always wanted to have a function that translates between Voigt and symmetric-rank-4 notation. Now I know where to look for.

bangerth commented:
Just a couple of comments.

I think I'm curious to do a before-after run to make sure the assembly hasn't gotten significantly slower. But I have no doubt about your claim that it hasn't.

jperryhouts commented:
OK, I incorporated your comments except the one about multiplication vs comparison. The multiplication was what slowed the code down before, and the comparison method fixed it. Let me know if you have another suggestion, because I agree it's a bit hard to read as is.

bangerth commented:
I'm merging this in the current state. May I ask you for a favor, though: could you create a follow-up patch that implements the "always multiply" strategy and create a pull-request from it? That would make it simpler to compare the two approaches side-by-side for others as well. My suspicion is that the multiplication of a rank-4 and a rank-2 tensor are simply not implemented very efficiently and that we could gain a good deal by simply optimizing the function that does so.

jperryhouts commented:
I'm sure it could be optimized. The difference between release mode, and debug mode was already a factor of 5, and there's probably more that could be done. But like you mentioned, that's just a lot of multiplications.
